### PR TITLE
Climbing system

### PR DESCRIPTION
**Pull request by nuke-makes-games**
Entities with a ClimbMode component can now click drag themselves onto anything with a Climbable component. Players can also drag other players onto tables, if the other player is incapacitated or dead.

### CODE REVIEW

**Comment by nuke-makes-games** [comment]
It passed all the checks, but i haven't been able to run tests locally, and i'd like to write a few integration tests so please don't merge yet please

File: Content.Client/GameObjects/Components/Movement/ClimbModeComponent.cs
Comment (Zumorica) [inline Content.Client/GameObjects/Components/Movement/ClimbModeComponent.cs:17]: I'm pretty sure you can point the TryGetComponent out parameter to your `_body` variable.

File: Content.Server/GameObjects/Components/Movement/ClimbModeComponent.cs
Comment (Zumorica) [inline Content.Server/GameObjects/Components/Movement/ClimbModeComponent.cs:74]: Why are you casting this? The return type should already be `TileRef`.

File: Content.Server/GameObjects/Components/Movement/ClimbModeComponent.cs
Comment (Zumorica) [inline Content.Server/GameObjects/Components/Movement/ClimbModeComponent.cs:42]: You could just make a property for this.

File: Content.Server/GameObjects/Components/Movement/ClimbModeComponent.cs
Comment (Zumorica) [inline Content.Server/GameObjects/Components/Movement/ClimbModeComponent.cs:40]: Make this a property.

File: Content.Server/GameObjects/Components/Movement/ClimbModeComponent.cs
Comment (Zumorica) [inline Content.Server/GameObjects/Components/Movement/ClimbModeComponent.cs:22]: Same here. Point the out parameter to your `body` field.

File: Content.Server/GameObjects/Components/Movement/ClimbModeComponent.cs
Comment (Zumorica) [inline Content.Server/GameObjects/Components/Movement/ClimbModeComponent.cs:13]: Just a nitpick: A better name in my opinion would be `ClimbingComponent`

File: Content.Server/GameObjects/Components/Movement/ClimbableComponent.cs
Comment (Zumorica) [inline Content.Server/GameObjects/Components/Movement/ClimbableComponent.cs:69]: Could this ever be null?

File: Content.Server/GameObjects/Components/Movement/ClimbableComponent.cs
Comment (Zumorica) [inline Content.Server/GameObjects/Components/Movement/ClimbableComponent.cs:197]: Is "them" what you wanted to use here? Also sucks because "He jumps" "She jumps" "They jump" the verb changes depending on the subject. This isn't your fault though, as I don't think we have a macro for this yet.

Comment (Zumorica) [reply Content.Server/GameObjects/Components/Movement/ClimbableComponent.cs:197]: Also you put `the` when you should probably use `{1:theName}` again.

File: Content.Server/GameObjects/Components/Movement/ClimbableComponent.cs
Comment (nuke-makes-games) [reply Content.Server/GameObjects/Components/Movement/ClimbableComponent.cs:69]: I wasn't sure so i added the sanity check. In theory you should never be dragging to a null entity though so maybe it's overkill.

File: Content.Server/GameObjects/Components/Movement/ClimbModeComponent.cs
Comment (nuke-makes-games) [reply Content.Server/GameObjects/Components/Movement/ClimbModeComponent.cs:13]: Yeah i struggled to figure out the proper name. Components for verbs maybe need a standardized way to name them, i think [Verb]ingComponent is probably good. Could even make it more verbose to delineate that it's a verb, it could be like [Verb]VerbComponent

I'll leave it to everyone else

**Comment by nuke-makes-games** [comment]
Alright, i made those changes, plus i added DoAfter functionality to the climbing. Assuming all changes made are 👍 then this should be ready for a merge.

I will be writing integration tests but first i want to tidy up the integration tests project a bit

**Comment by nuke-makes-games** [comment]
oof integration testing this is harder than i thought, a lot of the climb functionality is dependent on user input. my integration test is really barebones, probably not even worth adding. 😭

File: Content.Server/GameObjects/Components/Movement/ClimbableComponent.cs
Comment (DrSmugleaf) [reply Content.Server/GameObjects/Components/Movement/ClimbableComponent.cs]: This still needs to be resolved with `"{0:them} jumps onto {1:theName}!"`

File: Content.Server/GameObjects/Components/Movement/ClimbableComponent.cs
Comment (DrSmugleaf) [inline Content.Server/GameObjects/Components/Movement/ClimbableComponent.cs:164]: Vector2 should be imported instead of using the full import name.

File: Content.Shared/GameObjects/Components/Movement/SharedClimbingComponent.cs
Comment (DrSmugleaf) [inline Content.Shared/GameObjects/Components/Movement/SharedClimbingComponent.cs:10]: This name of this net id should be changed to `CLIMBING` as the name of the component was changed in https://github.com/space-wizards/space-station-14/pull/1750/commits/ae4fc15bcc39df8e3eff8e4e42cdda5131551133

File: Content.Shared/GameObjects/Components/Movement/SharedClimbingComponent.cs
Comment (DrSmugleaf) [inline Content.Shared/GameObjects/Components/Movement/SharedClimbingComponent.cs:39]: CollisionLayer has a collection of flags, meaning that this check would fail whenever the collided body has more collision layers than just vault impassable. This would be fixed by checking if the collided body and the owner share any collision layer flags, excluding vaultimpassable.

File: Content.Shared/Physics/ClimbController.cs
Comment (DrSmugleaf) [inline Content.Shared/Physics/ClimbController.cs:54]: Same as the change on line 61, to make the check more lenient.

```suggestion
            if (ControlledComponent.Owner.Transform.WorldPosition.EqualsApprox(_lastKnownPosition, 0.05))
```

**Review by DrSmugleaf — CHANGES_REQUESTED** [review]

File: Content.Shared/Physics/ClimbController.cs
Comment (DrSmugleaf) [inline Content.Shared/Physics/ClimbController.cs:61]: As we found out on Discord, these are percentages.
Oops.

```suggestion
            if (ControlledComponent.Owner.Transform.WorldPosition.EqualsApprox(_lastKnownPosition, 0.01))
            {
                _numTicksBlocked++;
            }

            _lastKnownPosition = ControlledComponent.Owner.Transform.WorldPosition;

            if (ControlledComponent.Owner.Transform.WorldPosition.EqualsApprox(_movingTo.Value, 0.01))
```

**Review by DrSmugleaf — CHANGES_REQUESTED** [review]

**Review by DrSmugleaf — APPROVED** [review]

**Pull request merged**
